library/src/amd_detail/hipsolver_refactor.cpp

```diff
@@ -0,0 +1,952 @@
+/* ************************************************************************
+ * Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell cop-
+ * ies of the Software, and to permit persons to whom the Software is furnished
+ * to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
+ * PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
+ * CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ *
+ * ************************************************************************ */
+
+/*! \file
+ *  \brief Implementation of the compatibility refactor APIs that require especial
+ *  calls to hipSOLVER or rocSOLVER.
+ */
+
+#include "error_macros.hpp"
+#include "exceptions.hpp"
+#include "hipsolver.h"
+#include "hipsolver_conversions.hpp"
+
+#include "rocblas/internal/rocblas_device_malloc.hpp"
+#include "rocblas/rocblas.h"
+#include "rocsolver/rocsolver.h"
+#include <algorithm>
+#include <climits>
+#include <functional>
+#include <iostream>
+#include <math.h>
+
+extern "C" {
+
+/******************** HANDLE ********************/
+struct hipsolverRfHandle
+{
+    hipsolverRfResetValuesFastMode_t fast_mode;
+    hipsolverRfMatrixFormat_t        matrix_format;
+    hipsolverRfUnitDiagonal_t        diag_format;
+    hipsolverRfNumericBoostReport_t  numeric_boost;
+
+    hipsolverRfFactorization_t   fact_alg;
+    hipsolverRfTriangularSolve_t solve_alg;
+
+    rocblas_handle   handle;
+    rocsolver_rfinfo rfinfo;
+
+    rocblas_int n, nnzA, nnzL, nnzU, nnzLU, batch_count;
+    double      effective_zero;
+    double      boost_val;
+
+    rocblas_int* dPtrA;
+    rocblas_int* dIndA;
+    double*      dValA;
+
+    rocblas_int *dPtrL, *hPtrL;
+    rocblas_int *dIndL, *hIndL;
+    double *     dValL, *hValL;
+
+    rocblas_int *dPtrU, *hPtrU;
+    rocblas_int *dIndU, *hIndU;
+    double *     dValU, *hValU;
+
+    rocblas_int *dPtrLU, *hPtrLU;
+    rocblas_int *dIndLU, *hIndLU;
+    double *     dValLU, *hValLU;
+
+    rocblas_int* dP;
+    rocblas_int* dQ;
+
+    uint8_t *d_buffer, *h_buffer;
```

Comment:
A buffer should be `void*` (or `char*` / `unsigned char*`). In practice, `uint8_t` is pretty much always a typedef for `unsigned char` and when that is the case, it inherits all the special exceptions to the aliasing rules that `char` gets, but that is not guaranteed.

Reply:
I seem to remember hearing once that I couldn't trust `char` to be 8 bits in length on all compilers/architectures, so I thought this might be safer since it's explicit. I'm happy to change it if you believe it's safe.

Reply:
That's a good thought, but it doesn't actually matter. These buffers are allocated by malloc / hipMalloc, so the size you request is denominated in bytes. The type pointed to by `{d,h}_buffer` is never used during the size calculation.

---
It's a bit of a tangent, but I might as well explain why `uint8_t` would have the opposite effect of what you're expecting.

If you used `new uint8_t[]` rather than malloc, then the size of the array type would be part of the calculation of the allocation size, but that would actually make your size calculation incorrect. On some esoteric platform where `CHAR_BIT == 4`, then `sizeof(uint8_t) == 2`. The sizes of the elements are calculated with `sizeof(T) * count`, so you'd end up allocating double the required memory by creating an array of `uint8_5[sizeof(T) * count]`. And, you'd have the opposite problem on a platform where `CHAR_BIT > 8`... but then `uint8_t` wouldn't exist on that platform so the program would not compile anyway.

This is a long way to say: you never need to worry about `CHAR_BIT` when doing these sorts of allocation size calculations. `char` is one byte by definition, and all calculations are done in bytes so it doesn't matter how big a byte is.